# Reference 
* https://ddanggle.gitbooks.io/interpy-kr/content/ch3-generators.html
* http://schoolofweb.net/blog/posts/%ED%8C%8C%EC%9D%B4%EC%8D%AC-%EC%A0%9C%EB%84%88%EB%A0%88%EC%9D%B4%ED%84%B0-generator/



# Iterable
* iterable은 iterator를 제공할 수 있는 모든 객체

# Iterator
* 이터레이터는 ```next(python2)``` 혹은 ```__next__```메소드가 정의 된 모든 객체를 말하는 것

# Generators
* 제너레이터는 이터레이터지만 단 한번만 반복함.
* 메모리에 모든 값을 저장하지 않기 때문에 값을 사용할 때 즉시 생성함.
* 'for' loop를 사용하거나 반복하는 함수나 구조에 생성된 값들을 전달하여 반복을 통해 사용
* 대부분의 제너레이터들은 함수로 구현 됨.
* 그러나, 값을 **return(반환)** 하지 않고, **yield(산출)**됨
* 제너레이터는 모든 결과물들을 메모리에 저장하지 않으면서 동시에, 많은 양의 결과 셋을 계산해야 할 때 최고(특히 루프 그 자체를 포함한 계산을 할 때)

In [1]:
def generator_function():
    for i in range(10):
        yield(i)

for item in generator_function():
    print(item)

0
1
2
3
4
5
6
7
8
9


In [2]:
def fibonacci(n):
    a = b = 1
    for i in range(n):
        yield a
        a, b = b, a+b

In [3]:
%%time
for x in fibonacci(10):
    print(x)

1
1
2
3
5
8
13
21
34
55
Wall time: 0 ns


### Generator Comprehension
* listcomprehension에서는 [] 괄호를 쓰지만 
* generator comprehension에서는 () 괄호를 씀

In [4]:
my_nums = (x * x for x in range(1, 10, 2))
print(my_nums)

<generator object <genexpr> at 0x000001E551C55468>


In [5]:
print(next(my_nums))
print(next(my_nums))
print(next(my_nums))
print(next(my_nums))
print(next(my_nums))

1
9
25
49
81


* 그 다음 next(my_nums)를 출력해보면?

In [6]:
print(next(my_nums))

StopIteration: 

* StopIteration 발생, 더 이상 전달할 값이 없다는 뜻
* 제너레이터는 일반적으로 for 루프를 통해서 사용함.
* for loop를 통해서 사용해보자

In [7]:
my_nums = (x * x for x in range(1, 10, 2))
print(my_nums)

<generator object <genexpr> at 0x000001E551CC4B48>


* for loop를 사용했을 때는 자신이 어디서 멈춰야 하는지 알고 있기 떄문에 StopIteration 발생하지 않음.

In [8]:
for num in my_nums:
    print(num)

1
9
25
49
81


* for loop를 사용하지 않고 한번에 제너레이터 데이터를 보고 싶으면 어떻게 해야 할까?
    * 그럴때는 제너레이터를 간단히 리스트로 변환하면 됨.
    * 그러나 한 번 리스트로 변형하면 제너레이터가 가지고 있던 장점을 모두 잃게 됨.

In [9]:
my_nums = (x * x for x in range(1, 10, 2))
print(my_nums)
print(list(my_nums))

<generator object <genexpr> at 0x000001E551CC4E60>
[1, 9, 25, 49, 81]


* 예제를 통해서 제너레이터의 퍼포먼스 확인

In [10]:
import os
import psutil
import random
import time

In [11]:
names = ['철수', '영희', '민호', '세훈', '민우']
majors = ['컴공', '국문학', '수학', '영문학', '통계학']

process = psutil.Process(os.getpid())
mem_before = process.memory_info().rss / 1024 / 1024

def people_list(num_people):
    result = []
    for i in range(num_people):
        person = {
            'id' : i,
            'name' : random.choice(names),
            'major' : random.choice(majors)
        }
        result.append(person)
    return result

def people_generator(num_people):
    for i in range(num_people):
        person = {
            'id': i,
            'name' : random.choice(names),
            'major' : random.choice(majors)
        }
        yield person
t1 = time.clock()
people = people_list(1000000)
t2 = time.clock()
mem_after = process.memory_info().rss / 1024 / 1024
total_time = t2 - t1

print('시작 전 메모리 사용량: {} MB'.format(mem_before))
print('시작 후 메모리 사용량: {} MB'.format(mem_after))
print('총 소요된 시간: {:.6f} 초'.format(total_time))

시작 전 메모리 사용량: 54.69921875 MB
시작 후 메모리 사용량: 387.45703125 MB
총 소요된 시간: 2.492141 초


* people_list(1000000)를 호출하여 백만명의 학생의 정보가 들어가는 리스트를 만들었을 때 메모리 사용량이 54MB에서 387MB로 늘었으며 시간은 약 2.49초가 걸렸음

In [13]:
names = ['철수', '영희', '민호', '세훈', '민우']
majors = ['컴공', '국문학', '수학', '영문학', '통계학']

process = psutil.Process(os.getpid())
mem_before = process.memory_info().rss / 1024 / 1024

def people_list(num_people):
    result = []
    for i in range(num_people):
        person = {
            'id' : i,
            'name' : random.choice(names),
            'major' : random.choice(majors)
        }
        result.append(person)
    return result

def people_generator(num_people):
    for i in range(num_people):
        person = {
            'id': i,
            'name' : random.choice(names),
            'major' : random.choice(majors)
        }
        yield person
t1 = time.clock()
people = people_generator(1000000)
for person in people:
    pass
t2 = time.clock()
mem_after = process.memory_info().rss / 1024 / 1024
total_time = t2 - t1

print('시작 전 메모리 사용량: {} MB'.format(mem_before))
print('시작 후 메모리 사용량: {} MB'.format(mem_after))
print('총 소요된 시간: {:.6f} 초'.format(total_time))

시작 전 메모리 사용량: 56.05859375 MB
시작 후 메모리 사용량: 56.05859375 MB
총 소요된 시간: 2.259472 초


* people_generator(1000000)를 호출했을 때는  시작 전 후 메모리 사용량의 변화량이 없는 것을 확인할 수 있고 시간도 이전 실험보다 더 적게 나오는 것을 확인할 수 있다.

* **generator(제너레이터)**의 퍼포먼스가 더 좋은 것을 확인할 수 있다.